packages/cozy-realtime/src/CozyRealtime.js

```diff
@@ -1,360 +1,549 @@
-   * @param {String}  eventName Event to unsubscribe from
-   * @param {Function}  handler   Function to call when an event of the
-   * given type on the given doctype or document is received from stack.
+   * Unregister event handlers for cozy-client events
+   * @private
    */
-  unsubscribe(eventName, type, handlerOrId, handlerOrUndefined) {
-    const { handler, id } = getHandlerAndId(handlerOrId, handlerOrUndefined)
-    validateParameters(eventName, type, id, handler)
-    const key = generateKey(eventName, type, id)
-
-    this.removeListener(key, handler)
+  unsubscribeClientEvents() {
+    this.client.removeListener('login', this.onClientLogin)
+    this.client.removeListener('tokenRefreshed', this.onClientTokenRefreshed)
+    this.client.removeListener('logout', this.onClientLogout)
+  }
 
-    if (!this._haveEventHandler()) {
-      this._resetSocket()
+  /**
+   * When the cozy-client instance successfully login - re-authenticate
+   * @private
+   */
+  onClientLogin() {
+    logger.info('Received login from cozy-client')
+    if (this.isWebSocketOpen()) {
+      this.authenticate()
+      // send subscriptions again, permissions may have changed
+      this.sendSubscriptions()
     }
   }
 
   /**
-   * Unsubscribe all handlers and close socket
+   * When the cozy-client instance refresh its access token - re-authenticate
+   * @private
    */
-  unsubscribeAll() {
-    this._getEventKeys()
-      .map(key => this._events[key].length > 0 && key)
-      .filter(Boolean)
-      .forEach(key => (this._events[key] = []))
+  onClientTokenRefreshed() {
+    // In theory, we don't have to do anything, and there is no need to authenticate again.
+    // We are reconnecting anyways because if we have a really bad timing, we may have sent an
+    // expired token for authentication, and we need to reconnect if that happen
+    if (this.hasWebSocket()) {
+      this.revokeWebSocket()
+    }
+  }
 
-    this._resetSocket()
+  /**
+   * When the cozy-client instance logs out - throw the socket, don't reconnect yet (no authentication)
+   * @private
+   */
+  onClientLogout() {
+    if (this.hasWebSocket()) {
+      this.revokeWebSocket()
+    }
   }
 
-  async send(doctype, id, data) {
-    return this._cozyClient.stackClient.fetchJSON(
-      'POST',
-      `/realtime/${doctype}/${id}`,
-      {
-        data
-      }
-    )
+  /* * * * * * * * * * * * * * * */
+  /* * Global (browser) EVENTS * */
+  /* * * * * * * * * * * * * * * */
+
+  /**
+   * Subscribe to global events
+   * @private
+   */
+  subscribeGlobalEvents() {
+    if (isInBrowser && window.addEventListener) {
+      window.addEventListener('online', this.onOnline)
+      window.addEventListener('offline', this.onOffline)
+    }
   }
 
-  _getEventKeys() {
-    if (!this._events) return []
+  /**
+   * Unsubscribe to global events
+   * @private
+   */
+  unsubscribeGlobalEvents() {
+    if (isInBrowser && window.removeEventListener) {
```

Comment:
I would prefer to repeat `typeof window !== 'undefined'` instead of using isInBrowser here. In the case where the `isInBrowser` logic changes and uses another object, other than window, the code would be more robust (here it is almost a coincidence that isInBrowser uses `window`)

Reply:
We ended up with `hasBrowserContext`